toolchain/check/context.h

```diff
@@ -616,12 +632,46 @@ class Context {
 
   auto global_init() -> GlobalInit& { return global_init_; }
 
+  // Marks the start of a region of insts in a pattern context that might
+  // represent an expression or a pattern.
+  auto BeginSubpattern() -> void;
+
+  // Ends a region started by BeginSubpattern (in stack order), treating it as
+  // an expression with the given result, and returns the ID of the region. The
+  // region will not yet have any control-flow edges into or out of it.
+  auto EndSubpatternAsExpression(SemIR::InstId result_id) -> SemIR::RegionId;
+
+  // Ends a region started by BeginSubpattern (in stack order), asserting that
+  // it was empty.
+  auto EndSubpatternAsEmpty() -> void;
+
+  // TODO: Add EndSubpatternAsPattern, when needed.
+
+  // Inserts the given region into the current code block. If the region
+  // consists of a single block, this will be implemented as a `splice_block`
+  // inst. Otherwise, this will end the current block with a branch to the entry
+  // block of the region, and add future insts to a new block which is the
+  // immediate successor of the region's exit block. As a result, this cannot be
+  // called more than once for the same region.
+  auto InsertHere(SemIR::RegionId region_id) -> SemIR::InstId;
```

Comment:
For now I think we should rename `Region` to `ExprRegion`, so I've done that. I've also restructured the code to avoid creating one to represent the whole function, because it doesn't even satisfy the single-exit invariant, much less the others that you allude to. Note also a subtle shift in terminology: the term "region" now means single-entry, but not necessarily single-exit.

While working on this I came across `Context::generic_region_stack`, which seems to use "region" in a lexical sense rather than the control-flow sense of `Context::region_stack`. That being the case, giving them such similar names seems liable to cause confusion. Do you have any thoughts about how to resolve that?

Reply:
That terminology sounds reasonable to me. Chandler might have views on using "region" to not mean "single-entry single-exit", might be worth checking whether that'll cause undue confusion.

I think the regions in `generic_region_stack` do at least roughly correspond to the regions in your sense. Eg, for a generic function we'll have one generic_region that roughly covers the decl's inst block (and that might eventually have control flow in it) and another that roughly covers the function body. Maybe not a precise match, but close enough?